File: typescript/0141-linked-list-cycle.ts
function hasCycle(head: ListNode | null): boolean {
    let slow = head;
    let fast = head;

    while (fast && fast.next) {
        if (slow) {
            slow = slow.next;
        }
        fast = fast.next.next;
        if (slow == fast) {
            return true;
        }
    }

    return false;
}

File: swift/0141-linked-list-cycle.swift
class Solution {
    func hasCycle(_ head: ListNode?) -> Bool {
        if head === nil {
            return false
        }
        
        var slow = head, fast = head
        
        while fast != nil && fast?.next != nil {
            slow = slow?.next
            fast = fast?.next?.next
            
            if fast === slow {
                return true
            }
        }
        return false
    }
}